File: src/main/java/com/rtomyj/yugiohAPI/model/Stats/DatabaseStats.java
package com.rtomyj.yugiohAPI.model.Stats;


import com.rtomyj.yugiohAPI.controller.stats.StatsController;
import com.rtomyj.yugiohAPI.model.HateoasLinks;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.Getter;
import org.springframework.hateoas.RepresentationModel;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

@Getter
@Builder
@ApiModel(
        description = "High level stats for data in the database."
        , parent = RepresentationModel.class
)
public class DatabaseStats extends RepresentationModel<DatabaseStats> implements HateoasLinks
{

    @ApiModelProperty(
            value = "Total number of products in the database."
            , accessMode = ApiModelProperty.AccessMode.READ_ONLY
    )
    private int productTotal;

    @ApiModelProperty(
            value = "Total number of cards in the database."
            , accessMode = ApiModelProperty.AccessMode.READ_ONLY
    )
    private int cardTotal;

    @ApiModelProperty(
            value = "Total number of ban lists in the database."
            , accessMode = ApiModelProperty.AccessMode.READ_ONLY
    )
    private int banListTotal;

    @ApiModelProperty(
            value = "Total number of years spanned/covered by ban lists stored in database."
            , accessMode = ApiModelProperty.AccessMode.READ_ONLY
    )
    private int yearsOfBanListCoverage;

    private static final Class<StatsController> statsControllerClass = StatsController.class;


    @Override
    public void setSelfLink()
    {

        this.add(
                linkTo(methodOn(statsControllerClass).getDatabaseStats()).withSelfRel()
        );

    }


    @Override
    public void setLinks()
    {
        this.setSelfLink();
    }

}

File: src/main/kotlin/com/rtomyj/skc/model/stats/DatabaseStats.kt
package com.rtomyj.skc.model.stats

import com.rtomyj.skc.controller.stats.StatsController
import com.rtomyj.skc.model.HateoasLinks
import io.swagger.v3.oas.annotations.media.Schema
import org.springframework.hateoas.RepresentationModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder

@Schema(
	implementation = DatabaseStats::class,
	description = "High level stats for data in the database."
)
data class DatabaseStats(
	@Schema(
		implementation = Int::class,
		description = "Total number of products in the database."
	) val productTotal: Int,
	@Schema(
		implementation = Int::class,
		description = "Total number of cards in the database.",
	) val cardTotal: Int,
	@Schema(
		implementation = Int::class,
		description = "Total number of ban lists in the database.",
	) val banListTotal: Int,
	@Schema(
		implementation = Int::class,
		description = "Total number of years spanned/covered by ban lists stored in database.",
	) val yearsOfBanListCoverage: Int
) : RepresentationModel<DatabaseStats>(), HateoasLinks {

	companion object {
		private val statsControllerClass = StatsController::class.java
	}


	override fun setSelfLink() {
		this.add(
			WebMvcLinkBuilder
				.linkTo(
					WebMvcLinkBuilder.methodOn(statsControllerClass)
						.databaseStats()
				)
				.withSelfRel()
		)
	}


	override fun setLinks() {
		setSelfLink()
	}
}